src/base/cvc4_assert.h

```diff
@@ -237,82 +273,90 @@ class InternalErrorException : public AssertionException {
  * debug builds only; in debug builds, it handles all assertion
  * failures (even those that exist in non-debug builds).
  */
-void debugAssertionFailed(const AssertionException& thisException,
+__CVC4__noreturn void debugAssertionFailed(const AssertionException& thisException,
                           const char* lastException);
 
-// If we're currently handling an exception, print a warning instead;
-// otherwise std::terminate() is called by the runtime and we lose
-// details of the exception
-#define AlwaysAssert(cond, msg...)                                     \
-  do {                                                                 \
-    if (__builtin_expect((!(cond)), false)) {                          \
-      /* save the last assertion failure */                            \
-      ::CVC4::LastExceptionBuffer* buffer =                            \
-          ::CVC4::LastExceptionBuffer::getCurrent();                   \
-      const char* lastException =                                      \
-          (buffer == NULL) ? NULL : buffer->getContents();             \
-      ::CVC4::AssertionException exception(#cond, __PRETTY_FUNCTION__, \
-                                           __FILE__, __LINE__, ##msg); \
-      ::CVC4::debugAssertionFailed(exception, lastException);          \
-    }                                                                  \
-  } while (0)
-
-#else /* CVC4_DEBUG */
-// These simpler (but less useful) versions for non-debug builds fails
-// will terminate() if thrown during stack unwinding.
-#define AlwaysAssert(cond, msg...)                                           \
-  do {                                                                       \
-    if (__builtin_expect((!(cond)), false)) {                                \
-      throw ::CVC4::AssertionException(#cond, __PRETTY_FUNCTION__, __FILE__, \
-                                       __LINE__, ##msg);                     \
-    }                                                                        \
-  } while (0)
+/* need to dispatch to ::CVC4::debugAssertionFailed on assertionException */
+template <>
+class Thrower<::CVC4::AssertionException> {
```

Comment:
I am not very comfortable splitting this across two headers. I think this can be avoided.